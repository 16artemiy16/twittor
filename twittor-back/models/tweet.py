from db import db


class TweetModel(db.Model):
    __tablename__ = 'tweets'

    id = db.Column(db.Integer, primary_key=True)
    body = db.Column(db.String(300))

    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)

    def json(self):
        return {'id': self.id, 'body': self.body, 'user_id': self.user_id}

    def save_to_db(self):
        db.session.add(self)
        db.session.commit()
