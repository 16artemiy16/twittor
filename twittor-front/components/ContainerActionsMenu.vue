<script lang="ts">
import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';
import Vue from 'vue';

interface ActionI {
  icon: string;
  text: string;
  onClick?: () => void;
}

@Component
export default class ContainerActionsMenu extends Vue {
  @Prop({ type: Array, default: () => [] })
  readonly actions!: ActionI[];

  get hasActions(): boolean {
    return !!this.actions.length;
  }

  onActionClick(action: ActionI) {
    if (action.onClick) {
      action.onClick();
    }
  }
}
</script>

<template>
  <div style="position: relative">
    <slot />
    <v-menu v-if="hasActions" bottom left>
      <template v-slot:activator="{ on, attrs }">
        <v-btn
          class="actions-menu"
          text
          icon
          v-bind="attrs"
          v-on="on"
        >
          <v-icon>mdi-dots-horizontal</v-icon>
        </v-btn>
      </template>
      <v-list light>
        <v-list-item
          v-for="(action, index) in actions"
          :key="index"
          @click="onActionClick(action)"
        >
          <v-list-item-icon>
            <v-icon>{{ action.icon }}</v-icon>
          </v-list-item-icon>
          <v-list-item-content>
            <v-list-item-title>{{ action.text }}</v-list-item-title>
          </v-list-item-content>
        </v-list-item>
      </v-list>
    </v-menu>
  </div>
</template>

<style scoped lang="scss">
.actions-menu {
  position: absolute;
  right: .5rem;
  top: .5rem;
}

.v-list-item {
  &:hover {
    background: $background-hover-pale;
  }
}

.v-list-item__icon {
  padding-right: 1rem;
}
</style>
