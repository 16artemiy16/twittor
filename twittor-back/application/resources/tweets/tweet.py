from flask_restful import Resource, reqparse
from flask_jwt_extended import jwt_required, get_jwt_identity
from flasgger import swag_from

from application.models import UserModel, TweetModel
from application.reqparsers import non_empty_string
from .utils import add_likes_details_empty


class Tweet(Resource):
    def __parse_post(self):
        parser = reqparse.RequestParser()
        parser.add_argument('body', required=True, type=non_empty_string, help='Body is required')
        return parser.parse_args()

    @jwt_required()
    @swag_from('tweet_post.yaml')
    def post(self):
        data = self.__parse_post()

        tweet = TweetModel(data['body'])
        tweet.user_id = get_jwt_identity()
        tweet.save_to_db()

        return {'tweet': add_likes_details_empty(tweet.json())}

    @jwt_required()
    @swag_from('tweet_delete.yaml')
    def delete(self, tweet_id):
        tweet = TweetModel.find_by_id(tweet_id)

        if not tweet:
            return {'message': 'There is no tweet with this id'}, 404

        user = UserModel.find_by_id(get_jwt_identity())

        if user.id != tweet.user_id:
            return {'message': 'You can\'t delete this tweet'}, 400

        tweet.delete_from_db()

        return {'message': 'The tweet have been successfully removed'}
