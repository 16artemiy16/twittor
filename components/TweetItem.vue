<script lang="ts">
import Component from 'vue-class-component';
import Vue from 'vue';
import { Prop } from 'vue-property-decorator';
import ContainerActionsMenu from '~/components/ContainerActionsMenu.vue';
import { TweetI } from '~/interfaces/tweet.interface';

interface ActionBottomI {
  icon: string;
  title: string;
  isActive?: string;
  text?: string;
}

@Component({
  components: {
    ContainerActionsMenu,
  },
})
export default class TweetItem extends Vue {
  @Prop({ type: Object })
  tweet!: TweetI;

  @Prop({ type: Boolean, default: false })
  skeleton!: boolean;

  readonly moreActions = [
    { icon: 'mdi-account-plus', text: 'Follow this User' },
    { icon: 'mdi-flag-outline', text: 'Report this Tweet' },
  ];

  readonly footerActions: ActionBottomI[] = [
    { icon: 'mdi-message-reply', title: 'Reply' },
    { icon: 'mdi-repeat-variant', title: 'Repost' },
    { icon: 'mdi-heart-outline', title: 'Like', text: '15' },
    { icon: 'mdi-export-variant', title: 'Share' },
  ];
}
</script>

<template>
  <ContainerActionsMenu :actions="!skeleton ? moreActions : []">
    <article class="tweet" :class="{ skeleton }">
      <div class="tweet__left">
        <img v-if="!skeleton" class="user-img" :src="tweet.user.img" width="35" height="35">
        <div v-else class="user-img"></div>
      </div>
      <div class="tweet__right">
        <div class="tweet__info">
          <template v-if="!skeleton">
            <span class="user-name">{{ tweet.user.name }}</span>
            <span class="user-login">{{ tweet.user.login }}</span>
            <span class="posted-date">3h</span>
          </template>
          <template v-else>&nbsp;</template>
        </div>
        <div class="tweet__body">
          <template v-if="!skeleton">
            {{ tweet.body }}
          </template>
          <template v-else>
            <p>&nbsp</p>
            <p>&nbsp</p>
          </template>
        </div>
        <div class="tweet__actions">
          <template v-if="!skeleton">
            <v-btn
              v-for="action in footerActions"
              :key="action.icon"
              text
              icon
            >
              <template>
                <v-icon>{{ action.icon }}</v-icon>
                {{ action.text  }}
              </template>
            </v-btn>
          </template>
          <template v-else>
            <span class="skeleton-action" v-for="action in footerActions"></span>
          </template>
        </div>
      </div>
    </article>
  </ContainerActionsMenu>
</template>

<style scoped lang="scss">
@keyframes shimmer {
  100% {
    transform: translateX(100%);
  }
}

.skeleton-action {
  padding: 1rem;
  border-radius: 20%;
  background: $skeleton;
  margin-top: .25rem;
}
.tweet {
  display: flex;
  padding: 1rem;
  cursor: pointer;
  width: 100%;
  position: relative;

  &.skeleton {
    &::after {
      position: absolute;
      top: 0;
      right: 0;
      bottom: 0;
      left: 0;
      transform: translateX(-100%);
      background-image: linear-gradient(
          90deg,
          rgba(#fff, 0) 0,
          rgba(#fff, 0.2) 20%,
          rgba(#fff, 0.5) 60%,
          rgba(#fff, 0)
      );
      animation: shimmer 2s infinite;
      content: '';
    }
    .user-img {
      width: 35px;
      height: 35px;
      background: $skeleton;
    }
    .tweet__info {
      width: 300px;
      background: $skeleton;
    }
    .tweet__body {
      p {
        background: $skeleton;
        &:not(:last-child) {
          margin-bottom: 0.25rem;
        }
      }
    }
  }

  &:hover {
    background: $background-hover-pale;
  }

  &__left {
    flex-basis: 50px;

    .user-img {
      border-radius: 50%;
    }
  }

  &__info {
    margin-bottom: .5rem;

    > * {
      margin-right: .5rem;
    }
  }

  &__right {
    flex-direction: column;
    flex: 1;
  }

  &__info {
    display: flex;

    .user-name {
      font-weight: bold;
    }
  }

  &__actions {
    display: flex;
    justify-content: space-between;
    padding-right: 5rem;
  }
}
</style>
