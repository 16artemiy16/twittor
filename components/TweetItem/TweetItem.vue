<script lang="ts">
import Component from 'vue-class-component';
import Vue from 'vue';
import { Emit, Prop } from 'vue-property-decorator';
import ContainerActionsMenu from '~/components/ContainerActionsMenu.vue';
import { TweetI } from '~/interfaces/tweet.interface';

interface ActionBottomI {
  icon: string;
  title: string;
  handler?: () => any;
  id?: string;
  text?: () => string;
}

@Component({
  components: {
    ContainerActionsMenu,
  },
})
export default class TweetItem extends Vue {
  @Prop({ type: Object, required: true })
  tweet!: TweetI;

  readonly moreActions = [
    { icon: 'mdi-account-plus', text: 'Follow this User' },
    { icon: 'mdi-flag-outline', text: 'Report this Tweet' },
  ];

  readonly footerActions: ActionBottomI[] = [
    { icon: 'mdi-message-reply', title: 'Reply' },
    { icon: 'mdi-repeat-variant', title: 'Repost' },
    {
      icon: 'mdi-heart-outline',
      title: 'Like',
      text: () => `${this.tweet.likes.total}`,
      id: 'like',
      handler: () => this.toggleLike(),
    },
    { icon: 'mdi-export-variant', title: 'Share' },
  ];

  @Emit('toggle-like')
  toggleLike() {
    const { id, likes } = this.tweet;
    return { id, isLike: !likes.isLikedByMe };
  }
}
</script>

<template>
  <ContainerActionsMenu :actions="moreActions">
    <article class="tweet">
      <div class="tweet__left">
        <img class="user-img" :src="tweet.user.img" width="35" height="35">
      </div>
      <div class="tweet__right">
        <div class="tweet__info">
          <template>
            <span class="user-name">{{ tweet.user.name }}</span>
            <span class="user-login">{{ tweet.user.login }}</span>
            <span class="posted-date">3h</span>
          </template>
        </div>
        <div class="tweet__body">
          <template>
            {{ tweet.body }}
          </template>
        </div>
        <div class="tweet__actions">
          <template>
            <v-btn
              class="tweet__action"
              :class="{ tweet__action_active: action.id === 'like' && tweet.likes.isLikedByMe }"
              @click="action.handler && action.handler(tweet)"
              v-for="action in footerActions"
              :key="action.icon"
              text
              icon
            >
              <template>
                <v-icon>{{ action.icon }}</v-icon>
                {{ action.text && action.text(tweet)  }}
              </template>
            </v-btn>
          </template>
        </div>
      </div>
    </article>
  </ContainerActionsMenu>
</template>

<style scoped lang="scss">
.tweet {
  display: flex;
  padding: 1rem;
  cursor: pointer;
  width: 100%;
  position: relative;

  &:hover {
    background: $background-hover-pale;
  }

  &__left {
    flex-basis: 50px;

    .user-img {
      border-radius: 50%;
    }
  }

  &__info {
    margin-bottom: .5rem;

    > * {
      margin-right: .5rem;
    }
  }

  &__right {
    flex-direction: column;
    flex: 1;
  }

  &__info {
    display: flex;

    .user-name {
      font-weight: bold;
    }
  }

  &__actions {
    display: flex;
    justify-content: space-between;
    padding-right: 5rem;
  }

  &__action {
    &_active {
      color: $like !important;
    }
  }
}
</style>
