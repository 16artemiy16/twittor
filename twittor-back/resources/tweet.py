from flask_restful import Resource, reqparse
from flask_jwt_extended import jwt_required, get_jwt_identity

from models.user import UserModel
from models.tweet import TweetModel
from models.like import LikeModel
from reqparsers.pagination import parse_pagination


def _add_likes_details(data, user_id):
    data['likes'] = {
        'total': LikeModel.get_count_in_tweet(data['id']),
        'isLikedByMe': LikeModel.is_tweet_liked_by_user(data['id'], user_id)
    }
    return data


class TweetListByUser(Resource):
    @jwt_required()
    def get(self, id):
        pagination = parse_pagination(reqparse.RequestParser())

        user = UserModel.find_by_id(id)

        if not user:
            return {'message': 'User with this id is not found.'}, 404

        paginated_tweets = user.tweets.offset(pagination['skip']).limit(pagination['limit'])
        total = user.tweets.count()
        has_next_page = total - (pagination['page'] * pagination['limit']) > 0

        return {
            'tweets': [_add_likes_details(t.json(), get_jwt_identity()) for t in paginated_tweets],
            'pagination': {
                'page': pagination['page'],
                'limit': pagination['limit'],
                'total': total,
                'hasNextPage': has_next_page
            }
        }


class Tweet(Resource):
    def __parse_post(self):
        parser = reqparse.RequestParser()
        parser.add_argument('body', required=True, help='Body is required', type=str)
        return parser.parse_args()

    @jwt_required()
    def post(self):
        data = self.__parse_post()

        tweet = TweetModel(data['body'])
        tweet.user_id = get_jwt_identity()
        tweet.save_to_db()

        return {'tweet': tweet.json()}


class ExploreTweets(Resource):
    # TODO: this is mocked for now to return the user's tweets
    @jwt_required()
    def get(self):
        pagination = parse_pagination(reqparse.RequestParser())

        user = UserModel.find_by_id(get_jwt_identity())

        paginated_tweets = user.tweets.offset(pagination['skip']).limit(pagination['limit'])
        total = user.tweets.count()
        has_next_page = total - (pagination['page'] * pagination['limit']) > 0

        return {
            'tweets': [_add_likes_details(t.json(), get_jwt_identity()) for t in paginated_tweets],
            'pagination': {
                'page': pagination['page'],
                'limit': pagination['limit'],
                'total': total,
                'hasNextPage': has_next_page
            }
        }
