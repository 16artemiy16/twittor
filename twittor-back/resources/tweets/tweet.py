from flask_restful import Resource, reqparse
from flask_jwt_extended import jwt_required, get_jwt_identity

from models.user import UserModel
from models.tweet import TweetModel
from reqparsers.types import non_empty_string


class Tweet(Resource):
    def __parse_post(self):
        parser = reqparse.RequestParser()
        parser.add_argument('body', required=True, type=non_empty_string, help='Body is required')
        return parser.parse_args()

    @jwt_required()
    def post(self):
        data = self.__parse_post()

        tweet = TweetModel(data['body'])
        tweet.user_id = get_jwt_identity()
        tweet.save_to_db()

        return {'tweet': tweet.json()}

    @jwt_required()
    def delete(self, tweet_id):
        tweet = TweetModel.find_by_id(tweet_id)

        if not tweet:
            return {'message': 'There is no tweet with this id'}, 404

        user = UserModel.find_by_id(get_jwt_identity())

        if user.id != tweet.user_id:
            return {'message': 'You can\'t delete this tweet'}, 400

        tweet.delete_from_db()

        return {'message': 'The tweet have been successfully removed'}
