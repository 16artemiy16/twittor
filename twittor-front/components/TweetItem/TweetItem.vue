<script lang="ts">
import Component from 'vue-class-component';
import Vue from 'vue';
import { Emit, Prop } from 'vue-property-decorator';
import ContainerActionsMenu from '~/components/ContainerActionsMenu.vue';
import { TweetI } from '~/interfaces/tweet.interface';
import { stringifyTimestampDiff } from '~/utils/date-diff.util';

interface ActionBottomI {
  icon: string;
  title: string;
  handler?: () => any;
  id?: string;
  text?: () => string;
}

@Component({
  components: {
    ContainerActionsMenu,
  },
})
export default class TweetItem extends Vue {
  @Prop({ type: Object, required: true })
  tweet!: TweetI;

  readonly moreActions = [
    { icon: 'mdi-account-plus', text: 'Follow this User' },
    { icon: 'mdi-flag-outline', text: 'Report this Tweet' },
  ];

  readonly footerActions: ActionBottomI[] = [
    { icon: 'mdi-message-reply', title: 'Reply' },
    { icon: 'mdi-repeat-variant', title: 'Repost' },
    {
      icon: 'mdi-heart-outline',
      title: 'Like',
      text: () => `${this.tweet.likes.total}`,
      id: 'like',
      handler: () => this.toggleLike(),
    },
    { icon: 'mdi-export-variant', title: 'Share' },
  ];

  get dateDiff(): string {
    return stringifyTimestampDiff(this.tweet.created);
  }

  @Emit('toggle-like')
  toggleLike() {
    const { id, likes } = this.tweet;
    return { id, isLike: !likes.isLikedByMe };
  }
}
</script>

<template>
  <ContainerActionsMenu :actions="moreActions">
    <article class="tweet d-flex pos-relative pa-4">
      <div class="tweet__left">
        <img class="user-img rounded-circle" :src="tweet.user.img" width="35" height="35">
      </div>
      <div class="flex-column flex-grow-1">
        <div class="tweet__info mb-2 d-flex">
          <template>
            <span class="user-name mr-2 font-weight-bold">{{ tweet.user.fullname }}</span>
            <span class="user-login mr-2">{{ tweet.user.login }}</span>
            <span class="posted-date">{{ dateDiff }}</span>
          </template>
        </div>
        <div class="tweet__body">
          {{ tweet.body }}
        </div>
        <div class="tweet__actions d-flex pr-16 justify-space-between">
          <template>
            <v-btn
              class="tweet__action"
              :class="{ tweet__action_active: action.id === 'like' && tweet.likes.isLikedByMe }"
              @click="action.handler && action.handler(tweet)"
              v-for="action in footerActions"
              :key="action.icon"
              text
              icon
            >
              <template>
                <v-icon>{{ action.icon }}</v-icon>
                {{ action.text && action.text(tweet)  }}
              </template>
            </v-btn>
          </template>
        </div>
      </div>
    </article>
  </ContainerActionsMenu>
</template>

<style scoped lang="scss">
.tweet {
  cursor: pointer;
  width: 100%;

  &:hover {
    background: $background-hover-pale;
  }

  &__left {
    flex-basis: 50px;
  }

  &__action {
    &_active {
      color: $like !important;
    }
  }
}
</style>
