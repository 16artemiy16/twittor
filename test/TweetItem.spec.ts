import { shallowMount, Wrapper } from '@vue/test-utils';
import { TweetI } from '~/interfaces/tweet.interface';
import TweetItem from '~/components/TweetItem/TweetItem.vue';
import ContainerActionsMenu from '~/components/ContainerActionsMenu.vue';

const TWEET: TweetI = {
  id: 'a',
  body: 'Hello! Welcome here!',
  user: {
    name: 'John',
    login: 'john',
    img: 'http://john.com/my-img.png',
  },
  likes: {
    total: 15,
    isLikedByMe: false,
  },
}

describe('TweetItem', () => {
  let wrapper!: Wrapper<TweetItem>;

  beforeEach(() => {
    wrapper = shallowMount(TweetItem, {
      propsData: {
        tweet: TWEET,
      },
      components: {
        ContainerActionsMenu,
      }
    });
  });

  it('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  });

  describe('should render text data correctly', () => {
    it('User name', () => {
      expect(wrapper.get('.user-name').text()).toBe(TWEET.user.name);
    })
  })
  it('test', () => {
    expect(true).toBe(true);
  });
});
