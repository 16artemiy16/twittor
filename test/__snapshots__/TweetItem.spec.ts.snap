// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TweetItem renders correctly 1`] = `
<containeractionsmenu-stub
  actions="[object Object],[object Object]"
>
  <article
    class="tweet"
  >
    <div
      class="tweet__left"
    >
      <img
        class="user-img"
        height="35"
        src="http://john.com/my-img.png"
        width="35"
      />
    </div>
     
    <div
      class="tweet__right"
    >
      <div
        class="tweet__info"
      >
        <span
          class="user-name"
        >
          John
        </span>
         
        <span
          class="user-login"
        >
          john
        </span>
         
        <span
          class="posted-date"
        >
          3h
        </span>
      </div>
       
      <div
        class="tweet__body"
      >
        
          Hello! Welcome here!
        
      </div>
       
      <div
        class="tweet__actions"
      >
        <v-btn-stub
          activeclass=""
          class="tweet__action"
          icon="true"
          tag="button"
          text="true"
          type="button"
        >
          <v-icon-stub>
            mdi-message-reply
          </v-icon-stub>
          
              
            
        </v-btn-stub>
        <v-btn-stub
          activeclass=""
          class="tweet__action"
          icon="true"
          tag="button"
          text="true"
          type="button"
        >
          <v-icon-stub>
            mdi-repeat-variant
          </v-icon-stub>
          
              
            
        </v-btn-stub>
        <v-btn-stub
          activeclass=""
          class="tweet__action"
          icon="true"
          tag="button"
          text="true"
          type="button"
        >
          <v-icon-stub>
            mdi-heart-outline
          </v-icon-stub>
          
              15
            
        </v-btn-stub>
        <v-btn-stub
          activeclass=""
          class="tweet__action"
          icon="true"
          tag="button"
          text="true"
          type="button"
        >
          <v-icon-stub>
            mdi-export-variant
          </v-icon-stub>
          
              
            
        </v-btn-stub>
      </div>
    </div>
  </article>
</containeractionsmenu-stub>
`;
